source/draw/gpu/opengl/backend/EglDisplayContext.ooc

```diff
@@ -0,0 +1,139 @@
+		This validate(eglDestroySurface(this _eglDisplay, this _eglSurface), EGL_TRUE, "eglDestroySurface")
+		This _mutex with(||
+			if (This _contextCount == 1)
+				This validate(eglTerminate(this _eglDisplay), EGL_TRUE, "eglTerminate")
+			This _contextCount -= 1
+		)
+		super()
+	}
+	getDisplay: func -> Pointer { this _eglDisplay }
+	printExtensions: func {
+		extensions := eglQueryString(this _eglDisplay, EGL_EXTENSIONS) as CString
+		extensionsString := String new(extensions, extensions length())
+		array := extensionsString split(' ')
+		extensionsString free()
+		Debug print("EGL Extensions: ")
+		for (i in 0 .. array count)
+			Debug print(array[i])
+		array free()
+	}
+	swapBuffers: func { eglSwapBuffers(this _eglDisplay, this _eglSurface) }
+	_chooseConfig: func (configAttribs: Int*) -> Pointer {
+		numConfigs: Int
+		eglChooseConfig(this _eglDisplay, configAttribs, null, 10, numConfigs&)
+		matchingConfigs: Pointer[numConfigs]
+		eglChooseConfig(this _eglDisplay, configAttribs, matchingConfigs[0]&, numConfigs, numConfigs&)
+		chosenConfig: Pointer = null
+
+		for (i in 0 .. numConfigs) {
+			success: UInt
+			red, green, blue, alpha: Int
+			success = eglGetConfigAttrib(this _eglDisplay, matchingConfigs[i], EGL_RED_SIZE, red&)
+			success &= eglGetConfigAttrib(this _eglDisplay, matchingConfigs[i], EGL_BLUE_SIZE, blue&)
+			success &= eglGetConfigAttrib(this _eglDisplay, matchingConfigs[i], EGL_GREEN_SIZE, green&)
+			success &= eglGetConfigAttrib(this _eglDisplay, matchingConfigs[i], EGL_ALPHA_SIZE, alpha&)
+
+			if (success && red == 8 && blue == 8 && green == 8 && alpha == 8) {
+				chosenConfig = matchingConfigs[i]
+				break
+			}
+		}
+		chosenConfig
+	}
+	_generateContext: func (shared, config: Pointer) -> Bool {
+		contextAttribs := [
+			EGL_CONTEXT_CLIENT_VERSION, 3,
+			EGL_NONE] as Int*
+		This _mutex with(|| This _contextCount += 1)
+		this _eglContext = eglCreateContext(this _eglDisplay, config, shared, contextAttribs)
+		if (this _eglContext == null) {
+			Debug print("Failed to create OpenGL ES 3 context, trying with OpenGL ES 2 instead")
+			contextAttribs = [
+				EGL_CONTEXT_CLIENT_VERSION, 2,
+				EGL_NONE] as Int*
+			this _eglContext = eglCreateContext(this _eglDisplay, config, shared, contextAttribs)
+			if (this _eglContext == null)
+				Debug error("Failed to create OpenGL ES 3 or OpenGL ES 2 context")
+			else
+				Debug print("WARNING: Using OpenGL ES 2")
+		}
+		This validate(eglMakeCurrent(this _eglDisplay, this _eglSurface, this _eglSurface, this _eglContext), EGL_TRUE, "eglMakeCurrent")
+		true
+	}
+	_initializeDisplay: func (display: Pointer) {
+		this _eglDisplay = eglGetDisplay(display)
+		This validate(this _eglDisplay != EGL_NO_DISPLAY, "eglGetDisplay")
+		This validate(eglInitialize(this _eglDisplay, null, null), EGL_TRUE, "eglInitialize")
+	}
+	_generate: func (display: Pointer, nativeBackend: Long, sharedContext: This) -> Bool {
```

Comment:
I tried overloads for `Gles3Context create` but in order to use them, I had to duplicate `GLContext createContext` to have any use of the new overloads. Default values for internal things like EGL might not be needed since it should be hidden inside the OpenGL context.